Sources/SemanticAnalyzer/TypeChecker.swift

```
@@ -123,24 +123,41 @@ public struct TypeChecker: ASTPass {
+      // Both sides must have type Int.
+      if ![lhsType, rhsType].contains(.errorType), !lhsType.isCompatible(with: .basicType(.int)) || !rhsType.isCompatible(with: .basicType(.int)) {
+        diagnostics.append(.incompatibleOperandTypes(operatorKind: binaryExpression.opToken, lhsType: lhsType, rhsType: rhsType, expectedTypes: [.basicType(.int)], expression: .binaryExpression(binaryExpression)))
+      }
+    case .and, .or:
+      // Both sides must have type Bool.
+      if ![lhsType, rhsType].contains(.errorType), !lhsType.isCompatible(with: .basicType(.bool)) || !rhsType.isCompatible(with: .basicType(.bool)) {
+        diagnostics.append(.incompatibleOperandTypes(operatorKind: binaryExpression.opToken, lhsType: lhsType, rhsType: rhsType, expectedTypes: [.basicType(.bool)], expression: .binaryExpression(binaryExpression)))
+      }
+    case .doubleEqual, .notEqual:
+      // Both sides must have the same type, and one of either Address, Bool, Int or String.
+      if ![lhsType, rhsType].contains(.errorType), !lhsType.isCompatible(with: rhsType) {
+        diagnostics.append(.unmatchedOperandTypes(operatorKind: binaryExpression.opToken, lhsType: lhsType, rhsType: rhsType, expression: .binaryExpression(binaryExpression)))
+      }
+      let acceptedTypes: [Type.RawType] = [.basicType(.address), .basicType(.bool), .basicType(.int), .basicType(.string)]
+      if ![lhsType, rhsType].contains(.errorType), !acceptedTypes.contains(lhsType) {
+        diagnostics.append(.incompatibleOperandTypes(operatorKind: binaryExpression.opToken, lhsType: lhsType, rhsType: rhsType, expectedTypes: acceptedTypes, expression: .binaryExpression(binaryExpression)))
+      }
+    case .openBrace, .closeBrace, .openSquareBracket, .closeSquareBracket, .colon, .doubleColon, .openBracket, .closeBracket, .arrow, .leftArrow, .comma, .semicolon, .doubleSlash, .dotdot, .ampersand, .halfOpenRange, .closedRange:
+      // These are not valid binary operators.
+      fatalError()
```

Would you not want to do default here?

---

I didn't want to add a default here so that when we add a new type of operator it will fail to compile and can be fixed rather than silently getting caught by the default.

---

Then I think moving these into a list is more important.